### ISSUE TITLE
Tutorial 5 trains with no validation set — published segments ship no `_validation_mask`

### ISSUE THREAD

**Issue by khj1222** [ScrollPrize/villa#1231]
Following [tutorial 5](https://scrollprize.org/tutorial5) end to end on `w00_20231016151002`
(native Windows, RTX 5090, `merge-ink-pipelines`), I ended up with a model I had no way to
evaluate, and I'd like to check whether that's intended.

### 1. Is the missing `_validation_mask` deliberate?

The pipeline clearly supports one:

- `data/patch_finding/default.py` marks every patch touching `_validation_mask` as a
  validation patch, and `_exclude_validation_voxels_from_training_supervision` in
  `data/ink_dataset.py` removes those voxels from the training supervision;
- `training/train.py` runs a validation pass every `val_every` steps and logs
  Confusion / BalancedAccuracy;
- `preprocessing/create_label_zarrs.py` already converts `_validation_mask.{tif,tiff,png}`;
- `evaluation/metrics/` implements DRD and a weighted pseudo-F-measure.

But the published segment carries only `_inklabels` and `_supervision_mask`, and the tutorial
never mentions validation. A stock tutorial run therefore gives:

```
flat_ink_patches_...json   "is_validation": true  ->     0
                           "is_validation": false -> 2,710
val_previews/              (empty)
```

`val_every` iterates an empty loader, and DRD / pseudo-F-measure never execute — so there is
no way to tell whether a change to the config, the model or the augmentations helped.

Are the validation masks kept internally and simply not published, or is generating one left
to the user? If the latter, is there a convention for choosing the held-out region that I
should match?

I ask because the choice turns out to matter more than I expected. On this segment the
supervision mask is **15 disconnected regions** around annotated letters (1.5%–20.7% of the
supervised area each, ink density 0.114–0.440), and 4 of them sit closer to another region
than one 256 px patch. A naive rectangular split cuts individual letters in half, so the
model trains on one stroke and is scored on the one beside it.

Holding out whole regions instead (20% of supervised area, ink density matched to 0.2283 vs
the segment's 0.2283) gives 1,337 validation / 2,240 training patches, and the numbers that
come out are:

| | best F1 | IoU |
|---|---|---|
| clean 20k run, scored on held-out regions | 0.8232 | 0.6995 |
| the pre-existing checkpoint that had trained on those regions | 0.8594 | 0.7535 |

Per-region F1 on the clean run spans 0.796–0.895, and a 3-fold rerun puts another fold at
0.8497 — i.e. the split matters by ~0.03 F1 on its own, which is larger than many changes
people would want to claim.

### 2. `create_label_zarrs` OOMs on striped TIFF input

Separate and much smaller, but it cost me a while: converting a full-resolution label image
only streams when the input TIFF is **tiled**. `_get_tiled_tiff_metadata` returns `None` for a
striped TIFF, `convert_image` falls back to `build_pyramid`, and that materializes the whole
65-deep volume:

```
ERROR ..._validation_mask.tif: Unable to allocate 25.1 GiB for an array
      with shape (65, 16125, 25690) and data type uint8
```

The shipped `_inklabels` / `_supervision_mask` TIFFs happen to be tiled 256×256, so this never
bites on published assets — but any label image written with a default `tifffile.imwrite` hits
it. Would you take a PR that either falls back to a streamed strip path, or fails early with a
clear "input must be tiled" message? Happy to open that as its own issue if you'd rather keep
this one to the validation question.

---

In the meantime I wrote the tooling I needed for the above — region-aware held-out mask
generation, evaluation against it (threshold sweep, your DRD / pseudo-F-measure classes, and a
per-region breakdown), a checkpoint sweep, and a k-fold driver. Happy to share or upstream it
in whatever shape you'd prefer, or to drop it if masks already exist on your side.

**Comment by khj1222** [comment]
Splitting a follow-up out of this thread rather than growing it, since it is a different claim:
the three ink_9um segments that *do* ship a `_validation_mask` turn out to split their annotation
within connected regions rather than by whole ones, and on pherc0139-w016 there is no leak-free
held-out subset available at all — 99.1% of its held-out pixels sit within two training patches
and the rest contains no ink. Numbers and a tool are in #1638.

This issue's question stands unchanged: whether the 26 segments that ship no mask are intended to
ship none.

**Comment by khj1222** [comment]
A narrower follow-up, and a question before I write anything rather than after.

I am not re-asking the first question in this issue — whether published segments are meant to ship a `_validation_mask` is still yours to answer, and I have stopped guessing at it. This is about the second half of the same gap: once you have a prediction, there is no supported way to score it.

**What I checked**, on `merge-ink-pipelines` at `3ea17f5`:

- `koine_machines/evaluation/metrics/` implements four metrics. Two of them — `Confusion` and `BalancedAccuracy` — are imported by `training/train.py` and run in the validation pass.
- The other two, `DRD` and `PFMWeighted`, have **no callers anywhere in the tree**: not in training, not in inference, not in tests. `grep -rn "DRD\|PFMWeighted"` outside their own two files returns nothing.
- `evaluation/__init__.py` and `evaluation/metrics/__init__.py` are both empty, and the package has no `__main__` and no console script, so there is no entry point either.

Put together: the validation metrics only ever run inside a training loop, over `validation_patches`, which is empty unless a segment ships a `_validation_mask` — and in the `ink_9um` corpus three segments do. A prediction TIFF sitting next to its labels cannot be scored by anything in this repository.

**The question.** Is an evaluation entry point something you want upstream, or is scoring deliberately left outside the pipeline? Either answer is useful to me and I will not push on it. If the answer is "not wanted", I would rather know now than send a PR that costs you a review to decline.

If it is wanted, two shapes, smallest first:

1. **Wire the two unused metrics into the existing validation pass.** Smaller than it sounds: both extend `BinaryImageMetric`, which already implements `compute_per_sample` over `_iter_sample_batches`, so `DRD(per_sample=True)` and `PFMWeighted(per_sample=True)` sit beside the two metrics the pass already builds, with no new plumbing. It does nothing for anyone without a validation split, but it stops two implemented metrics from being dead code.

   One caveat that argues against this being enough on its own: `BinaryImageMetric.threshold` defaults to 0.5, and on this data the threshold that maximises F1 is nowhere near it — across 126 scored cells my sweeps land between 68 and 153 out of 255, depending on the segment and the checkpoint. A DRD logged at a fixed 0.5 during training and a DRD reported from a sweep are not the same number, and I would not want them quoted as if they were.
2. **A `score` entry point**: `python -m koine_machines.evaluation.score <prediction> <segment_dir>`, sweeping the threshold and reporting the four metrics plus per-region breakdown as JSON. This is what I have been running outside the repo since July — the harness linked from this issue — so it is not speculative work, and upstreaming it means the numbers people quote come out of your code rather than mine.

I have no preference between them, and I am happy with neither. I have also not written 2 against your interfaces, deliberately: after #1638 I would rather ask than arrive with something built.

One thing I would need from you for either: whether region-level reporting belongs in it at all. My version breaks the score down per annotated region because that is what showed me a single split can move F1 by ~0.03 on this data, but that is my use, not necessarily yours.
